### PR TITLE
Added complexity regularization parameters (L1, L2, min_gain_to_split)

### PR DESCRIPTION
**Pull request by Allardvm**
These commits add three regularization parameters that discourage or constrain the complexity of the boosting trees. The parameters can reduce the degree of overfitting and have the added benefit of speeding up training because they limits the number of leaves. The parameters follow (as closely as possible) the implementation of XGBoost's regularization parameters (`alpha`, `lambda`, and `gamma`) to fit with users' expectations about appropriate values.

1. `lambda_l1 (default = 0.0f)` controls the L1 weights regularization by setting the strength of an L1 regularization term in the objective function (i.e. `lambda_l1 * |weights|`). This term is implemented by pushing `sum_gradients` towards zero in the weights calculations.
2. `lambda_l2 (default = 0.0f)` controls the L2 weights regularization by setting the strength of an L2 regularization term in the objective function (i.e. `0.5 * lambda_l2 * weights^2`, with the factor 0.5 mirroring XGBoost's implementation). This term is implemented by increasing `sum_hessians` in the weights calculations.
3. `min_gain_to_split (default = 0.0f)` sets the minimal gain / loss reduction required to split a leaf.

When used to train a binary classifier on a sizable (> 1 million data points) database, even somewhat naive and unoptimized values for these parameters improve performance beyond what could be realized with the current regularization parameters.

Nevertheless, I would appreciate it if you could check whether the current implementation seems performant and complete (i.e. was all code that calculates gain/loss reduction and output weights adjusted).

*Note that I did not yet update the example models with these parameters. It would be trivial to do this after we've settled on the appropriate parameter names and implementation.

### CODE REVIEW

File: src/treelearner/feature_histogram.hpp [hunk 1]
```diff
@@ -211,13 +213,28 @@ class FeatureHistogram {
...
+
+  /*!
   * \brief Calculate the split gain based on sum_gradients and sum_hessians
   * \param sum_gradients
   * \param sum_hessians
   * \return split gain
   */
   double GetLeafSplitGain(double sum_gradients, double sum_hessians) const {
-    return (sum_gradients * sum_gradients) / (sum_hessians);
+    if (lambda_l1_ != 0.0f) {
```
Comment (Allardvm) [inline src/treelearner/feature_histogram.hpp]: Yes, that's correct. The condition just prevents an unnecessary function call (and one or two additional `if`'s) when `lambda_l1_ == 0.0f`. The same goes for the`if` in `CalculateSplittedLeafOutput`. If you don't think this will make much (any?) difference in terms of performance, we should just get rid of this.

Comment (guolinke) [inline src/treelearner/feature_histogram.hpp]: The member function in .h/.hpp is ```inline``` automatically. So the cost of calling function is very small. 
So we can move this into ```L1RegularizedGradients``` , : ```if ( lambda_l1_ == 0.0f) { return sum_gradients; }```
However, it still have 3 ```if```'s conditions at worst case, can we reduce this ?

Comment (guolinke) [inline src/treelearner/feature_histogram.hpp]: BTW, sum_gradients_ is negative in most cases. So it will be better if exchanging Line221 and Line222.

Comment (Allardvm) [inline src/treelearner/feature_histogram.hpp]: I'll update pull with the changes proposed here. This'll also make it possible to immediately return `0.0f` from `GetLeafSplitGain()` and `CalculateSplittedLeafOutput()` when `!((sum_gradients > lambda_l1_) || (sum_gradients < lambda_l1_))`, thereby sidestepping quite a bit of computation.

As for reducing the number of `if`'s, we could reduce the calculations to:
```c++
if (abs(sum_gradients) > lambda_l1_) {
  return sign(sum_gradients) * (abs(sum_gradients) - lambda_l1_));
}
return 0.0f
```
A special version for `GetLeafSplitGain()` could then be a bit faster by omitting the multiplication with `sign(sum_gradients)`, because the output is squared afterwards anyway.

Nevertheless, it really comes down to how fast we can obtain the sign of `sum_gradients` in other cases and how fast a call to `std::abs` really is. There are some bit-shifts we can use for the sign, or it could be that the compiler knows a faster way to do `(x > 0) - (x < 0)`, but I will have to do some profiling to see if this is faster than the `if`-statements we have now.

File: src/treelearner/feature_histogram.hpp [hunk 2]
```diff
@@ -211,13 +213,28 @@ class FeatureHistogram {
 
 private:
   /*!
+  * \brief Return the L1 regularized sum_gradients
+  * \param sum_gradients
+  * \return regularized sum_gradients
+  */
+  double L1RegularizedGradients(double sum_gradients) const {
+  if (sum_gradients > lambda_l1_) return sum_gradients - lambda_l1_;
```
Comment (chivee) [inline src/treelearner/feature_histogram.hpp]: spaces

Comment (chivee) [inline src/treelearner/feature_histogram.hpp]: Making decision here can make code compatible and don't made further judgements

**Comment by Allardvm** [comment]
@chivee All three parameters should be fully functional—do you see any omissions?

@guolinke I have swapped the order of the conditions (negative `sum_gradients` is now first), got rid of the conditional `lambda_l1_=0.0f`, and have moved the L1 regularization logic into the `GetLeafSplitGain()` and `CalculateSplittedLeafOutput()` functions. Moving it here has the advantage that, in the case `NOT ((sum_gradients > lambda_l1_) || (sum_gradients < lambda_l1_))`, the regularization will zero out the `sum_gradients`, so the gain and output can be returned as `0.0f` immediately.

As for reducing the number of conditions (now just two): it is possible and my micro benchmarks suggest that it improves performance of these functions a little bit. I paste the alternative `GetLeafSplitGain()` and `CalculateSplittedLeafOutput()` functions in a discussion on these sections in the code.

File: src/treelearner/feature_histogram.hpp [hunk 3]
```diff
@@ -211,30 +213,48 @@ class FeatureHistogram {
...
   */
   double CalculateSplittedLeafOutput(double sum_gradients, double sum_hessians) const {
-    return -(sum_gradients) / (sum_hessians);
+    if (sum_gradients < -lambda_l1_) {
+      return -(sum_gradients + lambda_l1_) / (sum_hessians + lambda_l2_);
+    } else if (sum_gradients > lambda_l1_) {
+      return -(sum_gradients - lambda_l1_) / (sum_hessians + lambda_l2_);
+    }
+    return 0.0f;
   }
```
Comment (Allardvm) [inline src/treelearner/feature_histogram.hpp]: This seems to be slightly faster:
```c++
double CalculateSplittedLeafOutput(double sum_gradients, double sum_hessians) const {
  double abs_sum_gradients = fabs(sum_gradients);
  if (abs_sum_gradients > lambda_l1_) {
    return - copysign(abs_sum_gradients - lambda_l1_, sum_gradients) / (sum_hessians + lambda_l2_);
  }
  return 0.0f;
}
```

File: src/treelearner/feature_histogram.hpp [hunk 4]
```diff
@@ -211,30 +213,48 @@ class FeatureHistogram {
...
-    return (sum_gradients * sum_gradients) / (sum_hessians);
+    if (sum_gradients < -lambda_l1_) {
+      double reg_sum_gradients = sum_gradients + lambda_l1_;
+      return (reg_sum_gradients * reg_sum_gradients) / (sum_hessians + lambda_l2_);
+    } else if (sum_gradients > lambda_l1_) {
+      double reg_sum_gradients = sum_gradients - lambda_l1_;
+      return (reg_sum_gradients * reg_sum_gradients) / (sum_hessians + lambda_l2_);
+    }
+    return 0.0f;
   }
```
Comment (Allardvm) [inline src/treelearner/feature_histogram.hpp]: This seems to be slightly faster:
```c++
double GetLeafSplitGain(double sum_gradients, double sum_hessians) const {
  double abs_sum_gradients = fabs(sum_gradients);
  if (abs_sum_gradients > lambda_l1_) {
    double reg_abs_sum_gradients = abs_sum_gradients - lambda_l1_;
    return (reg_abs_sum_gradients * reg_abs_sum_gradients) / (sum_hessians + lambda_l2_);
  }
  return 0.0f;
}
```

File: src/treelearner/feature_histogram.hpp (hunk 3 above)
Comment (Allardvm) [inline src/treelearner/feature_histogram.hpp]: As far as I know, the compiler should output instructions that clear the sign bit so there shouldn't be any branching.

Comment (Allardvm) [inline src/treelearner/feature_histogram.hpp]: Looks like the assembly just uses `andps` to clear the sign bit, so that's pretty clean.

Comment (guolinke) [inline src/treelearner/feature_histogram.hpp]: Great. I think you can update to this version.

**Comment by Allardvm** [comment]
New version with less branching and slightly more arithmetic. Micro-benchmarks show that it is slightly faster.

File: src/treelearner/feature_histogram.hpp [hunk 5]
```diff
@@ -211,30 +213,45 @@ class FeatureHistogram {
...
   /*!
-  * \brief Calculate the output of a leaf based on sum_gradients and sum_hessians
+  * \brief Calculate the output of a leaf based on regularized sum_gradients and sum_hessians
   * \param sum_gradients
   * \param sum_hessians
   * \return leaf output
   */
   double CalculateSplittedLeafOutput(double sum_gradients, double sum_hessians) const {
-    return -(sum_gradients) / (sum_hessians);
+    double abs_sum_gradients = fabs(sum_gradients);
```
Comment (guolinke) [inline src/treelearner/feature_histogram.hpp]: use std::fabs

File: src/treelearner/feature_histogram.hpp [hunk 6]
```diff
@@ -211,30 +213,45 @@ class FeatureHistogram {
...
+    if (abs_sum_gradients > lambda_l1_) {
+      return -copysign(abs_sum_gradients - lambda_l1_, sum_gradients) / (sum_hessians + lambda_l2_);
```
Comment (guolinke) [inline src/treelearner/feature_histogram.hpp]: use std::copysign

File: src/treelearner/feature_histogram.hpp [hunk 7]
```diff
@@ -211,30 +213,45 @@ class FeatureHistogram {
...
   /*!
-  * \brief Calculate the split gain based on sum_gradients and sum_hessians
+  * \brief Calculate the split gain based on regularized sum_gradients and sum_hessians
   * \param sum_gradients
   * \param sum_hessians
   * \return split gain
   */
   double GetLeafSplitGain(double sum_gradients, double sum_hessians) const {
-    return (sum_gradients * sum_gradients) / (sum_hessians);
+    double abs_sum_gradients = fabs(sum_gradients);
```
Comment (guolinke) [inline src/treelearner/feature_histogram.hpp]: use std::fabs

Comment (Allardvm) [inline src/treelearner/feature_histogram.hpp]: Done. Note that all of this is defined in `<cmath>`, which is not explicitly included in `feature_histogram.hpp`, but included through `#include "split_info.hpp"`. This doesn't seem to be handled in a systematic way in this project, so do you have any preference on how to handle this?

**Comment by guolinke** [comment]
Nice work! Thanks again. 👍
